proposals/0002-rest-spread-operator.md

```diff
@@ -0,0 +1,133 @@
+# Feature name
+
+* Proposal: [HXP-0002](0002-rest-spread-operator.md)
+* Author: [Thomas Fétiveau](https://github.com/zabojad)
+
+
+## Introduction
+
+Add ES6-like `...` spread and rest operator to the Haxe syntax on both [arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator) and [objects](https://babeljs.io/docs/plugins/transform-object-rest-spread/).
+
+
+## Motivation
+
+Currently, there is a gap in Haxe for object duplication and overriding:
+
+* A lot of very popular js libs like React or Redux require developers to implement [pure functions](https://en.wikipedia.org/wiki/Pure_function) massively and this operator makes it really easy to do and avoid over-verbosity of simple operations.
+  
+  This:
+  ```
+  function pure(data : MyStruct) {
+
+  	return {...data, overwrittenFieldName: overwrittenFieldValue};
+  }
+  ```
+
+  is much less verbose than this:
+  ```
+  function pure(data : MyStruct) {
+
+  	var ret : MyStruct = {};
+
+  	for (f in Reflect.fields(data)) {
+
+  		Reflect.setField(ret,f,Reflect.field(data,f));
+  	}
+
+  	ret.overwrittenFieldName = overwrittenFieldValue;
+
+  	return ret;
+  }
+  ```
+
+* The standard javascript externs do not expose the [Object.assign API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign).
+
+  While it's not complicated to write an extern class for this, it feels wrong to add it to the source code of a custom project while it is something that should be available by default.
+
+* This `...` operator is part of the ECMAScript standard which Haxe syntax follows to some point. Getting closer to the ECMAScript syntax would ease the adoption of Haxe by javascript developers. 
+
+* Also, if added to the Haxe syntax, we would also get several other benefits out of it:
+```
+// array decomposition (non existent in Haxe)
+myFunction(...myArray);
+// shorter and more powerful array literals / comprehension
+var myNewArr = [ 1, 4, ...myOldArr, 56 ];
+// even augmented push
+myOldArr.push(...myNewArr);
+// object duplication
+var myNewObj = { ...myOldObj }
+// object duplication with field adding / overwriting
+var myNewObj = { ...myOldObj, overwrittenField: v, newField: y }
+```
+
+
+## Detailed design
+
+There are two cases to differentiate when using the `...` operator:
+
+* [On Objects](https://babeljs.io/docs/plugins/transform-object-rest-spread/):
+```
+// Spread properties
+var n = { x, y, ...z };
+trace(n); // { x: 1, y: 2, a: 3, b: 4 }
+
+// Rest properties
+var { x, y, ...z } = { x: 1, y: 2, a: 3, b: 4 };
+trace(x); // 1
+trace(y); // 2
+trace(z); // { a: 3, b: 4 }
+```
+
+We gain a powerful object duplication operator that is missing in Haxe right now.
+
+
+* On Arrays:
+
+```
+// Spread elements
+var myArr = [ 1, 2, 3 ];
+myFunc(...myArr);  // here, would do the same as: myFunc(1,2,3);
```

Comment:
These was described earlier, and it's quite clear what are they for. The "Detailed design" section is supposed to describe how it's going to work on the high level, for example:

* We need to introduce a new unary operator e.g. `OpSpread` that's parsed from `...<expr>` (which is actually a breaking change for macros, since adding new enum constructor will affect exhaustiveness checks).
* `{...obj, field2: 42}` will translate to `{field1: obj.field1, field2: 42}` (that's the idea, right?)
* `[1, 2, ...arr, 42, ...arr2, 1]` will translate to `var tmp = [1,2]; for (v in arr) tmp.push(v); tmp.push(42); for (v in arr2) tmp.push(v); tmp.push(1);`
* etc, etc.